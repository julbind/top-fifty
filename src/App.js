import React from 'react'
import './App.css'
import axios from 'axios'
import { useState } from 'react'
import _ from 'lodash'

function App() {
  const [artists, setArtists] = useState([])
  const getTopFiftyTracks = () => {
    return axios.get('https://ws.audioscrobbler.com/2.0/', {
      params: {
        method: 'chart.gettoptracks',
        api_key: process.env.REACT_APP_LASTFM_API_KEY,
        format: 'json'
      }
    })
    .then(response => {
      return response.data
    })
    .catch(error => {
      return error
    })
  }

  const mapTracksByArtist = () => {

    //helper func for constructing artistMap
    const getTrackObject = (track) => {
      return {
        name: track.name,
        playcount: track.playcount
      }
    }

    const artistMap = {}
    return getTopFiftyTracks().then(data => {
      data.tracks.track.forEach(track => {
        //would have preferred to use mbid, but some artists didn't have mbid
        const artistId = track.artist.url
        const artistEntry = artistMap[artistId]

        if (artistEntry) {
          artistEntry.tracks.push(getTrackObject(track))
        } else {
          artistMap[artistId] = {
            name: track.artist.name,
            tracks: [getTrackObject(track)]
          }
        }
      })
      return artistMap
    })
  }

  const getSortedArtists = () => {
    return mapTracksByArtist().then(artistMap => {
      if (!_.isEmpty(artistMap)) {
        //sort by number of tracks desc, then artist name asc
        return _.orderBy(Object.values(artistMap), [
          artist => artist.tracks.length,
          artist => artist.name],
          ['desc', 'asc']
        )
      }
      return []
    })
  }

  const getTopPortionOfArtists = (artistList, denom) => {
    if (!_.isEmpty(artists)) {
      //return top portion (as defined by denom) of artists, rounding up
      return artists.slice(0, Math.ceil(artists.length/denom))
    }
    return []
  }

  const getArtistsWithDivisiblePlaycounts = (artistList, denom) => {
    const newArtistList = []
    artistList.forEach(artist => {
      //add playcounts
      const totalPlaycount = artist.tracks.reduce((total, next) => {
        return total + parseInt(next.playcount)
      }, 0) //<-- default value
      //check if divisible by denom
      if (totalPlaycount % denom === 0) {
        newArtistList.push({
          name: artist.name,
          playcount: totalPlaycount,
        })
      }
    })
    return newArtistList
  }

  const getArtistTrackDisplay = (artist) => {
    if (artist.tracks) {
      const trackCount = artist.tracks.length
      return <div>
        {artist.name + ' — ' + trackCount +
        (trackCount === 1 ? ' song ' : ' songs ') +
        'in the top 50'}
      </div>
    }
  }

  const getArtistPlaycountDisplay = (artist) => {
    if (artist.playcount) {
      return <div>
        {artist.name + ' — ' + artist.playcount + ' plays'}
      </div>
    }
  }

  let artistDisplay = null
  if (_.isEmpty(artists)) {
    getSortedArtists().then(artists => setArtists(artists))
  } else {
    artistDisplay =
    <React.Fragment>
      <div className="artist-display" data-testid="artist-track-display">
        {artists.map(artist => (
          getArtistTrackDisplay(artist)
        ))}
      </div>
      <div className="artist-display" data-testid="artist-playcount-display">
        {getArtistsWithDivisiblePlaycounts(
          getTopPortionOfArtists(artists, 2), 9).map(artist => (
          getArtistPlaycountDisplay(artist)
        ))}
      </div>
    </React.Fragment>
  }


  return (
    <div className="App">
      <header className="App-header">
        { artistDisplay }
      </header>
    </div>
  );
}

export default App
