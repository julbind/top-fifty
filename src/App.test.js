import { render, screen } from '@testing-library/react'
import App from './App'
import axios from 'axios'
import { mockTopTracks } from './mockTopTracks'

jest.mock('axios')

describe('App', () => {
  beforeEach(() => {
    axios.get.mockResolvedValue(mockTopTracks)
    render(<App />)
  })
  it ('displays artists by count of tracks in top fifty', async () => {
    const el = await screen.findByTestId('artist-track-display')
    expect(el).toHaveTextContent(
      'Doja Cat — 5 songs in the top 50'
    )
  })
  it ('displays artists with playcounts divis by 9 in top half', async () => {
    const el = await screen.findByTestId('artist-playcount-display')
    expect(el).toHaveTextContent(
      'Kanye West — 21923361 plays'
    )
    expect(el).not.toHaveTextContent('Mitksi')
  })
})
