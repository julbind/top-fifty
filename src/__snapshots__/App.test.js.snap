// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App displays artists by count of tracks in top fifty 1`] = `undefined`;

exports[`App displays artists with playcounts divis by 9 in top half 1`] = `undefined`;
